[wasm] Use timezone abbreviations as fallback if full names don't exist

tqiu8 opened the pull request:
Fixes: #45136

Uses Timezone abbreviations as a fallback if full name does not exist in the ICU. Added 2 browser tests for daylight name and standard names.

tqiu8 committed:
use unix ver of GetDisplayName instead of invariant display name

tqiu8 commented:
@EgorBo Currently this doesn't work yet because `ucal_getTimeZoneDisplayName` from the icu doesn't return anything. Is it possible that the timezone display names have been filtered out? All standard, generic, and display names are being returned as empty even though the result code is successful.

jkotas commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -1872,9 +1872,9 @@
+    <Compile Include="$(MSBuildThisFileDirectory)System\TimeZoneInfo.GetDisplayName.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\IO\DriveInfoInternal.Browser.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\IO\PersistedFiles.Browser.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\TimeZoneInfo.GetDisplayName.Invariant.cs" />
```
Delete this file since it is no longer used? Also, `TimeZoneInfo.GetDisplayName.cs` can be merged back into `TimeZoneInfo.GetDisplayName.Unix.cs`.

radical commented:
This needs tests. And can you reference the issue, if any, that this is fixing?

lewing reviewed (commented):
It sounds like the invariant version was added when we didn't have full icu working.  If browser is the only one using it we should go ahead and remove it.  If you add a test we can verify it on desktop and compare to the wasm results.

tqiu8 committed:
Merge branch 'master' into tz-displayname-fix

tqiu8 committed:
Add tests for Daylight Names

tqiu8 committed:
Merge branch 'master' into tz-displayname-fix

tqiu8 committed:
Use tz abbreviations as fallback for browser cases, added test cases, fixed regex

tqiu8 committed:
Move abbrev fields to TimeZoneInfo.Unix.cs

radical committed:
[wasm] If standard, or daylight names are not available, then fallback

.. to abbreviations.

[This code](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.GetDisplayName.cs#L29-L40) seems to return `true` result, even though `timeZoneDisplayName` is null/empty. So, in such a case don't set the out var
and let the abbrev get used as the fallback (like it already says in the
comment).

radical commented:
The code here isn't completely correct. Eg. for Newfoundland, the display name would be `"(UTC-03:30) "`, instead of `"(UTC-03:30) NST"`.

I have reworked the tests a bit too. The final diff would be https://github.com/dotnet/runtime/compare/master...radical:rf-tz-displayname-fix , with my additional commit - https://github.com/radical/runtime/commit/1805c6d8889c6ff8254ed5f15ceac40282c8fb79

tqiu8 commented:
@radical The regex in the original form is actually incorrect. `"^\(UTC(\+|-)[0-9]{2}:[0-9]{2}\) \S.*"` is looking for a space after `(UTC-03:30)`.  Even though `((UTC-03:30)` may not contain any more text afterwards.

radical commented:
> @radical The regex in the original form is actually incorrect. `"^\(UTC(\+|-)[0-9]{2}:[0-9]{2}\) \S.*"` is looking for a space after `(UTC-03:30)`. Even though `((UTC-03:30)` may not contain any more text afterwards.

That was the case because the change in this PR set the display name (when falling back on abbreviation) *after* that string was constructed, so it ends up with nothing after the `(UTC+03:30) ` - https://github.com/dotnet/runtime/pull/45385/files#diff-08facbdd1840bc8f2298563a79fe4565de02c874d72d7ac944890f8c4d200659R87-R93

See my changes, and the test: https://github.com/radical/runtime/blob/rf-tz-displayname-fix/src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs#L90-L96

.. these have content after that space.

tqiu8 commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -1074,7 +1074,7 @@
       <Link>Common\Interop\Interop.TimeZoneDisplayNameType.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)Interop\Interop.TimeZoneInfo.cs" Condition="'$(TargetsBrowser)' != 'true'">
+    <Compile Include="$(CommonPath)Interop\Interop.TimeZoneInfo.cs">
```
Because we aren't using `TimeZoneInfo.GetDisplayName.Invariant.cs` anymore (will also remove that file in this PR) so `Interop.Globalization` is not defined for Browser

tqiu8 committed:
Merge pull request #4 from radical/rf-tz-displayname-fix

[wasm] If standard, or daylight names are not available, then fallback

lewing commented:
The change makes sense, but are there any tests that check the daylight name on desktop?

tqiu8 commented:
> The change makes sense, but are there any tests that check the daylight name on desktop?

Only this [one](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs#L2304), so nothing comprehensive

lewing reviewed (commented):
Let's add tests for the non browser case too then

tqiu8 committed:
add additional platform tests

tqiu8 committed:
edit displayname test

tqiu8 commented:
Test failures unrelated to this change.

lewing commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -81,6 +81,47 @@ public static void Names()
+
+        [Theory]
+        [MemberData(nameof(Platform_TimeZoneNamesTestData))]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
```
Does windows have names for these extra timezones?

mattjohnsonpint replied:
Yes, they would be:

```csharp
{ TimeZoneInfo.FindSystemTimeZoneById(s_strPacific), "(UTC-08:00) Pacific Time (US & Canada)", "Pacific Standard Time", "Pacific Daylight Time" },
{ TimeZoneInfo.FindSystemTimeZoneById(s_strSydney), "(UTC+10:00) Canberra, Melbourne, Sydney", "AUS Eastern Standard Time", "AUS Eastern Daylight Time" },
{ TimeZoneInfo.FindSystemTimeZoneById(s_strPerth), "(UTC+08:00) Perth", "W. Australia Standard Time", "W. Australia Daylight Time" },
{ TimeZoneInfo.FindSystemTimeZoneById(s_strIran), "(UTC+03:30) Tehran", "Iran Standard Time", "Iran Daylight Time" },
{ s_NewfoundlandTz, "(UTC-03:30) Newfoundland", "Newfoundland Standard Time", "Newfoundland Daylight Time" },
{ s_catamarcaTz, "(UTC-03:00) City of Buenos Aires", "Argentina Standard Time", "Argentina Daylight Time" }
```

lewing reviewed (approved):
Nice!

mattjohnsonpint commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -81,6 +81,47 @@ public static void Names()
+                    { TimeZoneInfo.FindSystemTimeZoneById(s_strPacific), "(UTC-08:00) Pacific Standard Time", "Pacific Standard Time", "Pacific Daylight Time" },
+                    { TimeZoneInfo.FindSystemTimeZoneById(s_strSydney), "(UTC+10:00) Australian Eastern Standard Time", "Australian Eastern Standard Time", "Australian Eastern Daylight Time" },
+                    { TimeZoneInfo.FindSystemTimeZoneById(s_strPerth), "(UTC+08:00) Australian Western Standard Time", "Australian Western Standard Time", "Australian Western Daylight Time" },
+                    { TimeZoneInfo.FindSystemTimeZoneById(s_strIran), "(UTC+03:30) +0330", "+0330", "+0430" },
```
Shouldn't these last three be the full names from ICU?

```csharp
{ TimeZoneInfo.FindSystemTimeZoneById(s_strIran), "(UTC+03:30) Iran Standard Time", "Iran Standard Time", "(UTC+03:30) Iran Standard Time" },
{ s_NewfoundlandTz, "(UTC-03:30) Newfoundland Standard Time", "Newfoundland Standard Time", "(UTC-03:30) Newfoundland Standard Time" },
{ s_catamarcaTz, "(UTC-03:00) Argentina Standard Time", "Argentina Standard Time", "(UTC-03:00) Argentina Standard Time" }
```

(Sorry I didn't see this earlier)

mattjohnsonpint replied:
How did the test pass with the wrong values?

tqiu8 replied:
Are you referring to the last value in each test? They're supposed to be the DaylightNames.

mattjohnsonpint replied:
No, I meant that in the non-browser version, the last three rows were not using the ICU names.  I believe its due to the `if (PlatformDetection.IsBrowser)` in the test - this part of the test isn't being run.  Per our offline discussion, I'll fix that up in my PR.

tarekgh commented:
@mattjohnsonpint if you can fix such issues in your PR would be great. You are touching the same area anyway.

mattjohnsonpint commented:
Will do.

lewing commented:
Thanks for following up on that, I should have caught it.
